plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,564 @@
+                    rb.status());
+        }
+      }
+    },
+    [&](const query::extract& extract) {
+      if (extract.policy == query::extract::preserve_ids) {
+        return self->send(extract.sink, table_slice{});
+      }
+      for (const auto& rb : record_batch_reader) {
+        if (rb.ok()) {
+          const auto slice = arrow_table_slice_builder::create(*rb);
+          auto final_slice = filter(slice, expr, ids);
+          if (final_slice) {
+            num_hits += final_slice->rows();
+            self->send(extract.sink, *final_slice);
+          } else {
+            // TODO: error handling?
+            fmt::print(stderr, "tp;no slice today\n");
+          }
+        }
+      }
+    },
+  };
+  caf::visit(handle_query, query.cmd);
+  return num_hits;
+}
+
+std::shared_ptr<arrow::Schema> parse_arrow_schema_from_metadata(
+  const std::shared_ptr<parquet::FileMetaData>& parquet_metadata) {
+  if (!parquet_metadata)
+    return {};
+  auto arrow_metadata
+    = parquet_metadata->key_value_metadata()->Get("ARROW:schema");
+  if (!arrow_metadata.ok())
+    return {};
+  auto decoded = detail::base64::decode(*arrow_metadata);
+  auto schema_buf = std::make_shared<arrow::Buffer>(decoded);
+  arrow::ipc::DictionaryMemo dict_memo;
+  arrow::io::BufferReader input(schema_buf);
+  auto arrow_schema = arrow::ipc::ReadSchema(&input, &dict_memo);
+  if (!arrow_schema.ok())
+    return {};
+  return *arrow_schema;
+}
+
+caf::expected<std::shared_ptr<arrow::Table>>
+read_parquet_buffer(const chunk_ptr& chunk) {
+  const auto* ptr
+    = static_cast<const uint8_t*>(static_cast<const void*>(chunk->data()));
+  auto bufr = std::make_shared<arrow::io::BufferReader>(
+    ptr, detail::narrow_cast<int64_t>(chunk->size()));
+  auto parquet_reader = parquet::ParquetFileReader::Open(bufr);
+  auto arrow_schema
+    = parse_arrow_schema_from_metadata(parquet_reader->metadata());
+  std::unique_ptr<parquet::arrow::FileReader> file_reader{};
+  if (!parquet::arrow::OpenFile(bufr, arrow::default_memory_pool(),
+                                &file_reader)
+         .ok())
+    return caf::exit_reason::unhandled_exception;
+  std::shared_ptr<arrow::Table> table{};
+  if (!file_reader->ReadTable(&table).ok())
+    return caf::exit_reason::unhandled_exception;
+  return map_enum_dictionary(arrow_schema, table);
+}
+
+system::store_actor::behavior_type
+store(system::store_actor::stateful_pointer<store_state> self,
+      const system::accountant_actor& accountant,
+      const system::filesystem_actor& fs, const uuid& id) {
+  self->state.self_ = self;
+  self->state.id_ = id;
+  self->state.accountant_ = accountant;
+  self->state.fs_ = fs;
+  self->state.path_ = store_path_for_partition(id);
+  self->request(self->state.fs_, caf::infinite, atom::mmap_v, self->state.path_)
+    .then(
+      [self](const chunk_ptr& chunk) {
+        if (auto table = read_parquet_buffer(chunk); table) {
+          self->state.table_ = *table;
```

Comment:
Can we assert here or inside the `read_parquet_buffer` function that the adjusted schema of the table matches the schema in the parquet metadata exactly?

Reply:
That's actually difficult because we reconstruct the table expicitly with the desired schema - I don't see what we could compare against here.

Two things:
1. `arrow_table_slice_builder::create(arrow::RecordBatch)` runs `Verify` which during development threw errors when the table schema and array type did not match
2. while traversing the schema tree, I added:
```c++
    [&](const auto&) -> std::optional<std::shared_ptr<arrow::ChunkedArray>> {
      VAST_ASSERT(t.to_arrow_type()->Equals(array->type()));
      return {};
    },
```
To make sure that the array type matches the schema type on every node and every level. Similar code has been added to the regular `arrow::Array` handlers.

I hope that addresses your concern.

Reply:
Yes, that works perfectly for me. Thanks!